set DX array type

orbeckst opened the pull request:
Issue #35 (and MDAnalysis/mdanalysis#1725) report that DX files written by gridDataFormats cannot be read by PyMOL because PyMOL (wrongly) *hard codes* the expected array type as "double". 

This PR adds a `type` kwarg to `Grid.export()` so that users now have the option to set the output data type to "double" if they so choose; the data are cast to the closest numpy.dtype.

By default, the numpy.dtype of the underlying array is used to set the closest matching DX data type.

kain88-de commented on gridData/OpenDX.py:
```diff
@@ -182,18 +192,118 @@ def write(self,file):
+        "float": "float32",       # default
+        "double": "float64",
+        # "string" not automatically supported
+    }
```
are they exact copies of each other only reverted? Wouldn't it be better to use `dx_types = {v: k for v,k in six.iteritems(np_types)}` so that we only have one list to take care of?

orbeckst replied:
No, they are not inverted, there's not a 1-to-1 correspondence.

kain88-de commented on gridData/core.py:
```diff
@@ -363,10 +381,11 @@ def _export_python(self, filename):
             cPickle.dump(data, f, cPickle.HIGHEST_PROTOCOL)
 
-    def _export_dx(self, filename):
+    def _export_dx(self, filename, **kwargs):
```
why not use `type=none` instead of undocumented `**kwargs`?

orbeckst replied:
changed

kain88-de commented on gridData/OpenDX.py:
```diff
@@ -182,18 +192,118 @@ def write(self,file):
+        ----------
+        classid : int
+        array : array_like
+        type : str (optional)
```
I would prefer to name this `dtype` for **D**ata**Type** following the numpy naming. That way we also avoid clashing with the buildin `type`.

orbeckst replied:
dtype is misleading because they are not numpy.dtypes. They are the specific labels in the DX file format specification. But "dxtype" seemed clunky and I didn't see a problem in shadowing the builtin type if it was clearer to the user. I also did not want to use dxtype as argument for export because if we ever want to use the data type for other formats then we can parse type appropriately.

kain88-de commented on gridData/OpenDX.py:
```diff
@@ -182,18 +192,118 @@ def write(self,file):
+        # "string" not automatically supported
+    }
+    dx_types = {
+        "byte": "uint8",
```
is unsigned the default in DX files?

orbeckst replied:
no, float

kain88-de replied:
Sorry I meant is a byte really automatically unsigned?

kain88-de replied:
for compatibility we should here map `char` to an `uint8`. 

https://web.archive.org/web/20080808140524/http://opendx.sdsc.edu/docs/html/pages/usrgu068.htm#Header_440

kain88-de commented on gridData/core.py:
```diff
@@ -363,10 +381,11 @@ def _export_python(self, filename):
             cPickle.dump(data, f, cPickle.HIGHEST_PROTOCOL)
 
-    def _export_dx(self, filename):
+    def _export_dx(self, filename, type=None, **kwargs):
```
kwargs can be removed now.

orbeckst replied:
Technically yes; practically it says that all `_export_*()` methods need to be able to take all kwargs and only process the ones that are needed. 

That's also the reason for `_export_python()` just using `**kwargs` and not doing anything with it.

kain88-de commented on gridData/core.py:
```diff
@@ -346,11 +346,29 @@ def export(self, filename, file_format=None):
+        exporter(filename, type=type)
 
-    def _export_python(self, filename):
+    def _export_python(self, filename, **kwargs):
```
where is `kwargs` used in the function? Wouldn't explicit named args be better?

orbeckst replied:
See https://github.com/MDAnalysis/GridDataFormats/pull/36#discussion_r157713756

I could add `type=None` and then ignore it but it seems clearer (to me, at least...) that `**kwargs` means that the method does not know what it will get and it will just take what it needs.

orbeckst replied:
I added an explanatory comment in cd3fdb051b86d3a2559461411cbcec60f98102d7

orbeckst commented:
I have to get some sleep... if you want to merge then just squash everything. Thanks for reviewing.

orbeckst commented:
Yes, that's my interpretation of the description of the allowed values on the page for the DX format ---see the link somewhere in the PR (sorry, can't get at it easily from mobile). 

> Am Dec 19, 2017 um 07:15 schrieb Max Linke <notifications@github.com>:
> 
> @kain88-de commented on this pull request.
> 
> Sorry I meant is a byte really automatically unsigned?
> 
> 
>

kain88-de commented on gridData/OpenDX.py:
```diff
@@ -182,18 +192,118 @@ def write(self,file):
+        "int": "int32",
+        "signed int": "int32",
+        "unsigned int": "uint64", # not explicit in DX, for compatibility
+        "int": "int64",           # not explicit in DX, for compatibility
```
int appears twice in this list. We can only map to a 32 or 64 bit one. Given the age of the format my guess is to use 32bit.

kain88-de commented on gridData/OpenDX.py:
```diff
@@ -182,18 +192,118 @@ def write(self,file):
+        "uint32": "unsigned int",
+        "int32": "int",           # DX "signed int"   equivalent
+        "uint64": "unsigned int", # not explicit in DX, for compatibility
+        "int64": "int",           # not explicit in DX, for compatibility
```
are they stored in binary or plain values? Do we have enough spaces to write the maximal/minimal values of this type into the dx file?

orbeckst commented:
I'm on holidays and won't do much coding over the next few days. If you want to finish the PR then that's super-cool, otherwise I see if I can get to it after Xmas. 

- yes , one int needs to be removed, the last one as you said 

- DX stores ascii text and I just write with str()... maybe there are better ways. 

--
Oliver Beckstein
email: orbeckst@gmail.com


> Am Dec 19, 2017 um 14:53 schrieb Max Linke <notifications@github.com>:
> 
> @kain88-de commented on this pull request.
> 
> Looks good besides some last questions about the type mapping.
> 
> In gridData/OpenDX.py:
> 
> > +        "float16": "float",       # float16 not available in DX, use float
> +        # numpy "float128 not available, raise error
> +        # "string" not automatically supported
> +    }
> +    dx_types = {
> +        "byte": "uint8",
> +        "unsigned byte": "uint8",
> +        "signed byte": "int8",
> +        "unsigned short": "uint16",
> +        "short": "int16",
> +        "signed short": "int16",
> +        "unsigned int": "uint32",
> +        "int": "int32",
> +        "signed int": "int32",
> +        "unsigned int": "uint64", # not explicit in DX, for compatibility
> +        "int": "int64",           # not explicit in DX, for compatibility
> int appears twice in this list. We can only map to a 32 or 64 bit one. Given the age of the format my guess is to use 32bit.
> 
> In gridData/OpenDX.py:
> 
> > +        "int8": "signed byte",
> +        "uint16": "unsigned short",
> +        "int16": "short",         # DX "signed short" equivalent
> +        "uint32": "unsigned int",
> +        "int32": "int",           # DX "signed int"   equivalent
> +        "uint64": "unsigned int", # not explicit in DX, for compatibility
> +        "int64": "int",           # not explicit in DX, for compatibility
> +        # "hyper",                # ?
> +        "float32": "float",       # default
> +        "float64": "double",
> +        "float16": "float",       # float16 not available in DX, use float
> +        # numpy "float128 not available, raise error
> +        # "string" not automatically supported
> +    }
> +    dx_types = {
> +        "byte": "uint8",
> for compatibility we should here map char to an uint8.
> 
> https://web.archive.org/web/20080808140524/http://opendx.sdsc.edu/docs/html/pages/usrgu068.htm#Header_440
> 
> In gridData/OpenDX.py:
> 
> > +    See `Array Objects`_ for details.
> +
> +    .. _Array Objects:
> +       https://web.archive.org/web/20080808140524/http://opendx.sdsc.edu/docs/html/pages/usrgu068.htm#Header_440
> +    """
> +    # equivalence between DX types and numpy dtypes.name
> +    # (round-tripping is not guaranteed to produce identical types)
> +    np_types = {
> +        "uint8": "byte",         # DX "unsigned byte" equivalent
> +        "int8": "signed byte",
> +        "uint16": "unsigned short",
> +        "int16": "short",         # DX "signed short" equivalent
> +        "uint32": "unsigned int",
> +        "int32": "int",           # DX "signed int"   equivalent
> +        "uint64": "unsigned int", # not explicit in DX, for compatibility
> +        "int64": "int",           # not explicit in DX, for compatibility
> are they stored in binary or plain values? Do we have enough spaces to write the maximal/minimal values of this type into the dx file?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

kain88-de commented:
Thanks @orbeckst

orbeckst commented:
I need to add docs (to show how things write files for pymol...) and I think I want to follow your earlier suggestion and not use type as a kwarg but instead use atype for "array type". 



> Am Dec 30, 2017 um 02:47 schrieb Max Linke <notifications@github.com>:
> 
> Thanks @orbeckst
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

orbeckst commented:
rebased, cleaned up history... ready to merge.

(Decided to leave kwarg `type` but added extensive docs on #35 which can also be linked in PR https://github.com/MDAnalysis/mdanalysis/pull/1746)
